File: Data Structures & Algorithms/three-integer-sum/submission-1.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {

        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);

        for (int idx = 0; idx < nums.length; idx++){

            if (nums[idx] > 0) return result;

            if (idx > 0 && nums[idx - 1] == nums[idx]) continue;

            int left = idx + 1, right = nums.length - 1;

            while (left < right){

                int curr = nums[idx] + nums[left] + nums[right];

                if (curr == 0){
                    result.add(Arrays.asList(nums[idx], nums[left], nums[right]));
                    left++;
                    right--;

                    while (left < right && nums[left - 1] == nums[left]){
                        left++;
                    }
                    while (left < right && nums[right + 1] == nums[right]){
                        right--;
                    }
                }
                else if (curr < 0){
                    left++;
                } else {right--;}
            }
        }

        return result;


        
    }
}

File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int left = 0, right = heights.size() - 1;

        int ans = 0;

        while (left < right){
            const int curr = (right - left) * (heights[left] < heights[right] ? heights[left] : heights[right]);
            ans = curr > ans ? curr: ans;

            if (heights[left] < heights[right]) left ++;
            else right--;
        }
        return ans;
    }
};
